### DISCUSSION TITLE
What's the correct way to deploy from another CI system such as circle ci?

### DISCUSSION THREAD

**Discussion by paymog** [vercel/vercel#7849] [Help] [answered]
I'm trying to migrate to vercel and I'm having a lot of trouble figuring out how to deploy from another CI system such as CircleCI.

I see that the `deploy` subcommand accepts a token which is super useful. However, I'm having trouble figuring out how to:

1. disable the auto-deployments by the vercel webapp which is connected to my repo and
2. correctly linking my CircleCI to the vercel project and scope

I see that `vercel link` might be useful but I see it asks for a scope but doesn't offer an option for setting a scope.

Here's me walking through the `vercel link command`:

```
 ❯❯❯ vercel link --token=<token> -p tripvector-mono
Vercel CLI 24.2.4
? Set up “~/code/tripvector/tripvector-mono/frontend”? [Y/n] y
? Which scope should contain your project?
● paymahn-protonmailc
○ Tripvector
```

and here's the help output:

```
 ❯❯❯ vercel link --help
Vercel CLI 24.2.4

  ▲ vercel link [options]

  Options:

    -h, --help                     Output usage information
    -A FILE, --local-config=FILE   Path to the local `vercel.json` file
    -Q DIR, --global-config=DIR    Path to the global `.vercel` directory
    -d, --debug                    Debug mode [off]
    -t TOKEN, --token=TOKEN        Login token
    -p NAME, --project=NAME        Project name
    --confirm                      Confirm default options and skip questions
```

I notice that when I try to deploy, I don't have an option for picking a project in addition to a scope:

```
 ❯❯❯ vercel deploy --help
Vercel CLI 24.2.4

  ▲ vercel [options] <command | path>

  Commands:

    Basic

      billing                          Manages the account payment methods
      deploy               [path]      Performs a deployment (default)
      dev                              Start a local development server
      env                              Manages the Environment Variables for your current Project
      init                 [example]   Initialize an example project
      ls | list            [app]       Lists deployments
      inspect              [id]        Displays information related to a deployment
      link                 [path]      Link local directory to a Vercel Project
      login                [email]     Logs into your account or creates a new one
      logout                           Logs out of your account
      pull                 [path]      Pull your Project Settings from the cloud
      switch               [scope]     Switches between teams and your personal account
      help                 [cmd]       Displays complete help for [cmd]

    Advanced

      rm | remove          [id]        Removes a deployment
      bisect                           Use binary search to find the deployment that introduced a bug
      domains              [name]      Manages your domain names
      projects                         Manages your Projects
      dns                  [name]      Manages your DNS records
      certs                [cmd]       Manages your SSL certificates
      secrets              [name]      Manages your global Secrets, for use in Environment Variables
      logs                 [url]       Displays the logs for a deployment
      teams                            Manages your teams
      whoami                           Shows the username of the currently logged in user

  Options:

    -h, --help                     Output usage information
    -v, --version                  Output the version number
    --cwd                          Current working directory
    -V, --platform-version         Set the platform version to deploy to
    -A FILE, --local-config=FILE   Path to the local `vercel.json` file
    -Q DIR, --global-config=DIR    Path to the global `.vercel` directory
    -d, --debug                    Debug mode [off]
    -f, --force                    Force a new deployment even if nothing has changed
    --with-cache                   Retain build cache when using "--force"
    -t TOKEN, --token=TOKEN        Login token
    -p, --public                   Deployment is public (`/_src` is exposed)
    -e, --env                      Include an env var during run time (e.g.: `-e KEY=value`). Can appear many times.
    -b, --build-env                Similar to `--env` but for build time only.
    -m, --meta                     Add metadata for the deployment (e.g.: `-m KEY=value`). Can appear many times.
    -C, --no-clipboard             Do not attempt to copy URL to clipboard
    -S, --scope                    Set a custom scope
    --regions                      Set default regions to enable the deployment on
    --prod                         Create a production deployment
    -c, --confirm                  Confirm default options and skip questions
```

I believe that my project is called `tripvector-mono` and my scope is called `tripvector` based on how this web page is structured
<img width="802" alt="image" src="https://user-images.githubusercontent.com/1986950/169666048-5fb7cb0d-a163-4f76-9b50-e01de29a7ae7.png">

and how these project settings are structured

<img width="987" alt="image" src="https://user-images.githubusercontent.com/1986950/169666059-ca0c0620-c79e-44d0-9537-d8bd64379d22.png">

What is the correct way to use the vercel CLI to deploy to `tripvector/tripvector-mono` from an external CI/CD system while also disabling automatic deployments from the Vercel webapp?

**Comment by paymog** [comment]
If the `deploy` command gave the option of picking a project as well, this would be much easier:

```
vercel deploy --prod --token=<token> --scope tripvector --project tripvector-mono
```

but it seems that's not available. Is it? Is there another way to non-interactively select both the project and scope with a token?

**Comment by barelyhuman** [comment]
1. Remove the integration from git (vercel dashboard)  
2. Generate a local manifest using `vercel link` , check the new manifest in `.vercel` directory. Add a simple bash script to generate the same manifest using secrets and environment variables on CircleCI
3. Use the token and `vercel deploy`
4. Let me know if it breaks somewhere

**Comment by leerob** [comment, marked as answer]
Hey everyone! :wave:

Happy to share it's now much easier to deploy to Vercel with any custom CI/CD provider like CircleCI.

[We've created a guide](https://vercel.com/support/articles/how-can-i-use-github-actions-with-vercel) and [an example](https://github.com/vercel/examples/tree/main/ci-cd/github-actions) using GitHub Actions and plan to continue creating further examples.

**Reply by lucashfreitas** [reply]
@leerob how does `vercel` "guess" what project it should use? I am tried to use an existent project but it created a new one - didn't see any option to set the project using the cli.

**Comment by Danubius1st** [comment]
I can deploy my nextjs app from localhost, using commands line.
But when I try to follow [your guide](https://vercel.com/support/articles/how-can-i-use-github-actions-with-vercel) I get this error:
![image](https://github.com/vercel/vercel/assets/22094979/bcfd4d71-4986-4910-82d4-82234d3fc5b7)
I checked the environment variables in the project settings in CircleCI:
![image](https://github.com/vercel/vercel/assets/22094979/20c00385-38de-4be2-bfc1-3da63b4edf28)
Is there something that only I miss?

**Reply by LeoAnt02** [reply]
Same issue here, any solution ?

**Reply by Danubius1st** [reply]
In a way: I gave up on vercel. github takes over all the work.

**Reply by Mahmoud-D** [reply]
I got same problem after follow this guide
